### PR TITLE
[DRAFT]: Add a Code Splitting page to the 'Recipes' section

### PR DESCRIPTION
**Pull request by abettadapur**
An initial draft at a code splitting page. @markerikson Let me know your thoughts, if I need to add more sections or clarify anything in more detail

### CODE REVIEW

**Comment by abettadapur** [comment]
#3186

File: docs/recipes/CodeSplitting.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+}
+```
+
+To add a new reducer, one can now call `store.reducerManager.add("asyncState", asyncReducer)`.
+
+To remove a reducer, one can now call `store.reducerManager.remove("asyncState")`
+
+## Libraries and Frameworks
+There are a few good libraries out there that can help you add the above functionality automatically:
+ * [`redux-dynamic-reducer`](https://github.com/ioof-holdings/redux-dynamic-reducer)
```
Comment (mpeyper) [inline docs/recipes/CodeSplitting.md]: Hey, I'm the author/maintainer of `redux-dynamic-reducer` and while I'm stoked you want to include my library in these docs, it's unfortunately out of active development and [will be deprecated soon](https://github.com/ioof-holdings/redux-dynamic-reducer/issues/16).  

Luckily, we do have a new library called [`redux-dynostore`](https://github.com/ioof-holdings/redux-dynostore) which supersedes it, adding more configurability and dynamic sagas support and you still get the `addReducers` function on the store.

**Comment by markerikson** [comment]
I'll try to take a look at this in the next few days - been busy with ReactConf and work.

File: docs/recipes/CodeSplitting.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
+# Code Splitting
+In large web applications, it is often desirable to split up the app code into multiple JS bundles that can be loaded on-demand. This strategy, called 'code splitting', helps to increase performance of your application by reducing the size of the initial JS payload that must be fetched.
+
+To code split with Redux, we want to be able to dynamically add reducers to the store. The default usage of Redux mandates that a single root reducer should to be passed to the `configureStore` call, which makes dynamically adding new reducers more difficult. Below, we discuss some approaches to solving this problem and reference two libraries that provide this functionality.
```
Comment (markerikson) [inline docs/recipes/CodeSplitting.md]: Rephrase this a bit.  It's not that "default usage mandates a single root reducer".  It's that there _is_ only one reducer, and the standard usage is to split that up into smaller functions for maintainability.

File: docs/recipes/CodeSplitting.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+## Basic Principle
+To dynamically add Reducers to a Redux store, there are two approaches that can be taken
```
Comment (markerikson) [inline docs/recipes/CodeSplitting.md]: I'd say there's only one approach - calling `store.replaceReducer()`.  However, there may be some different ways you can handle re-creating the root reducer.

File: docs/recipes/CodeSplitting.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+### Using `replaceReducer`
```
Comment (markerikson) [inline docs/recipes/CodeSplitting.md]: Let's have a small section on what `replaceReducer()` is here, then move any discussion of `injectReducers()` or similar to an additional header section.

File: docs/recipes/CodeSplitting.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+To add a new reducer, one can now call `store.reducerManager.add("asyncState", asyncReducer)`.
+
+To remove a reducer, one can now call `store.reducerManager.remove("asyncState")`
+
+## Libraries and Frameworks
+There are a few good libraries out there that can help you add the above functionality automatically:
+ * [`redux-dynamic-reducer`](https://github.com/ioof-holdings/redux-dynamic-reducer)
+    * This library exposes the `addReducer` function on the Redux store to accomplish the behavior we described above. It also has React bindings which make it easier to add reducers within the React component lifecycle.
+* [`redux-dynamic-modules`](https://github.com/Microsoft/redux-dynamic-modules)
+    * This library introduces the concept of a 'Redux Module', which is a bundle of Redux artifacts (reducers, middleware) that should be dynamically loaded. It also exposes a React higher-order component to load 'modules' when areas of the application come online. Additionally, it has integrations with libraries like `redux-thunk` and `redux-saga` which also help dynamically load their artifacts (thunks, sagas).
```
Comment (markerikson) [inline docs/recipes/CodeSplitting.md]: Might as well toss in a link to the list of additional libs I've seen: https://github.com/markerikson/redux-ecosystem-links/blob/master/reducers.md#dynamic-reducer-injection

**Comment by markerikson** [comment]
Just made those tweaks and merged this in.  @abettadapur , thanks for writing this!

**Comment by raRaRa** [comment]
It would be nice addition to know how to access the store in order to call `injectReducers`. E.g. if I were to write HOC or use it from some routes, how would I access `injectReducers`?

Some people recommend creating and then exporting the store, which other components or logic can import. What would be the recommended way? Could the docs show a small simple HOC in order to achieve this with a connected component?

**Comment by mpeyper** [comment]
Right now, the options are:

1. `export`/`import` the store directly (I would not recommend this approach myself)
2. Use the `store` out of the `ReactReduxContext` export of `react-redux` (not recommend as it is not a "public" API)
3. Create your own `Provider` and `Consumer` components to access the store/inject function
4. Use one of the libraries listed in the docs instead of rolling your own (they usually are using 2. internally but at least the unsafe access is hidden away from your code)

The problem is none of these are great solutions and I'd struggle to recommend any of them or want to endorse them by making them examples in the docs, except perhaps using a library, which I don't think there should be an example of in the `redux` docs.

See [this issue](https://github.com/reduxjs/react-redux/issues/1150) and [this PR](https://github.com/reduxjs/react-redux/pull/1154) for details and please leave feedback if you have any.

**Comment by mpeyper** [comment]
There's no built in way to do this and I'm not sure how I would a achieve this myself but I believe the [redux-dynamic-modules](https://github.com/Microsoft/redux-dynamic-modules) library supports this.

**Comment by raRaRa** [comment]
@mpeyper Yeah they do, but migrating to `redux-dynamic-modules` is not straightforward for everyone. For example, I would need to create a module for all external reducers that I'm currently using. I've opened an issue with proposal that could simplify it.

Thanks

**Pull request merged**